### ISSUE TITLE
Volume Mount hostpath: Unable to mount volumes for pod: timeout expired waiting for volumes to attach or mount

### ISSUE THREAD

**Issue by srinicrick65** [kubernetes/kubernetes#100790]
hi Team I am facing the below issue. Any solution for this ? I am blocked guys as there is a POC . I tried with k3d, Kind, minikube and its there everywhere.

k3d version v4.4.0
k3s version v1.20.4-k3s1 (default)
kubectl version : 1.2.0
helm : 3.0.2 (Using helm for deploying the pod)

Though these files and folders are there . Even if I remove type still the mount does not happen and it throws the same error
Error
gent-6c9494c59-p9xqm to kind-control-plane
3s 2s 3 edge-agent-6c9494c59-p9xqm Pod Warning FailedMount {kubelet kind-control-plane} MountVolume.SetUp failed for volume "socket-volume" : open /var/ability/rpc/socket: no such file or directory
3s 3s 1 edge-agent-6c9494c59 ReplicaSet Normal SuccessfulCreate {replicaset-controller } Created pod: edge-agent-6c9494c59-p9xqm
3m 3m 1 edge-agent-757bb75497-l8l6h Pod Normal Scheduled {default-scheduler } Successfully assigned edge-system/edge-agent-757bb75497-l8l6h to kind-control-plane
3m 2m 8 edge-agent-757bb75497-l8l6h Pod Warning FailedMount {kubelet kind-control-plane} MountVolume.SetUp failed for volume "cert-volume" : hostPath type check failed: /var/ability/certs is not a directory
3m 2m 8 edge-agent-757bb75497-l8l6h Pod Warning FailedMount {kubelet kind-control-plane} MountVolume.SetUp failed for volume "socket-volume" : hostPath type check failed: /var/ability/rpc/socket is not a socket file
3m 2m 8 edge-agent-757bb75497-l8l6h Pod Warning FailedMount {kubelet kind-control-plane} MountVolume.SetUp failed for volume "module-volume" : hostPath type check failed: /var/ability/modules is not a directory
1m 1m 1 edge-agent-757bb75497-l8l6h Pod Warning FailedMount {kubelet kind-control-plane} Unable to attach or mount volumes: unmounted volumes=[mqtt-secret-volume default-token-xjtdc module-volume edge-privatekey socket-volume edge-publickey mqtt-volume mqtt-auth-volume cert-volume], unattached volumes=[mqtt-secret-volume default-token-xjtdc module-volume edge-privatekey socket-volume edge-publickey mqtt-volume mqtt-auth-volume cert-volume]: timed out waiting for the condition

Please find my deployment.yaml file

```
apiVersion: apps/v1
kind: Deployment
metadata:
  name: edge-agent
  namespace: edge-system
  labels:
    moduleId: edge-agent
spec:
  replicas: 1
  selector:
    matchLabels:
      moduleId: edge-agent
  strategy:
    type: Recreate
  template:
    metadata:
      labels:
        moduleId: edge-agent
    spec:
      containers:
      - image: ability.azurecr.io/edge/agent:pocfeature_rebasemaster
        securityContext:
          privileged: true
          runAsUser: 0
        name: edge-agent
        imagePullPolicy: Always
        env:
        - name: DPS_PROVISIONING_HOST
          value: global.azure-devices-provisioning.net
        - name: DPS_ID_SCOPE
          value: 0ne00091CC4
        - name: DEVICE_ID
          value: 8fcb0635-9cc2-4f95-8914-0301668be636
        - name: CERT_FILE
          value: "/var/ability/certs/edgedevice-cert.pem"
        - name: KEY_FILE
          value: "/var/ability/certs/edgedevice-key.pem"
        - name: EDGE_TYPE
          value: "ability.device.edge.shpatil@1"
        - name: EDGEPROXYIMAGE
          value: "ability.azurecr.io/edge/proxy:2.4.12"
        - name: EDGEMQTTIMAGE
          value: "ability.azurecr.io/edge/broker:1.5.8"
        - name: MODULES_VOLUME
          value: "/var/ability/modules"
        - name: MQTT_VOLUME
          value: "/var/ability/mosquitto"
        - name: files_directory
          value: "/var/ability/modules"
        - name: ACL_ALLOWED_SERVER_NAMES
          value: "^.+\\.core\\.windows\\.net$ ^.+\\.azure-devices.net$ ^global\\.azure-devices-provisioning\\.net$"
        - name: SYS_TRAIL_AGGREGATOR_CA_PEM
          value: /etc/rsyslog-ca.pem
        - name: SYS_TRAIL_AGGREGATOR_HOST
          value: ability-edge-development-sys-log-dummy-certificate
        - name: SYS_TRAIL_AGGREGATOR_PORT
          value: "6514"
        - name: BLOCK_INCOMING_CONTACT_ON_IFACE
          value: dum8
        - name: LAYER2_CONNECTIVITY_GATEWAY_IP
          value: "203.0.113.1"
        - name: LAYER2_CONNECTIVITY_SUBNET_CIDR
          value: 203.0.113.0/24
        - name: LAYER2_CONNECTIVITY_CONTAINERS_CIDR
          value: 203.0.113.2/32
        - name: CONTAINER_BRIDGE_IP_CIDR
          value: 172.17.0.1/16
        - name: LAN_ACCESS_NAT_CIDR
          value: 172.19.0.0/16
        - name: INTERNET_ACCESS_NAT_CIDR
          value: 172.20.0.0/16
        - name: EDGENETWORK_EXT
          value: iot-internet
        - name: EDGENETWORK_INT
          value: iot-lan
        - name: EDGENETWORK_MACVLAN
          value: iot-layer2
        - name: LOGGER_LEVEL
          value: silly
        - name: SSL_ENGINE
          value: /var/ability/modules/libSecStoreEngine.so
        - name: DEV_QA_CERT_CHOICE
          value: CommonDevQAPlatform
        - name: HOST_RPC_SERVER_PATH
          value: /var/ability/rpc/socket


        ports:
        - containerPort: 80
          name: http
          protocol: TCP
        volumeMounts:
        - name: cert-volume
          mountPath: /var/ability/certs
        - name: module-volume
          mountPath: /var/ability/modules
        - name: socket-volume
          mountPath: /var/ability/rpc
        - name: mqtt-secret-volume
          mountPath: /var/run/secrets/mqtt/edge-agent
        - name: edge-privatekey
          mountPath: /var/run/secrets/pem/edge-agent-private-key
        - name: edge-publickey
          mountPath: /var/run/secrets/pem/edge-agent-public-key
        - name: mqtt-volume
          mountPath: /var/ability/mosquitto
        - name: mqtt-auth-volume
          mountPath: /var/ability/mqtt_auth


      volumes:
      - name: cert-volume
        hostPath:
          path: /var/ability/certs
          type: Directory
      - name: module-volume
        hostPath:
          path: /var/ability/modules
          type: Directory
      - name: socket-volume
        hostPath:
          path: /var/ability/rpc/socket
          type: Socket
      - name: mqtt-secret-volume
        hostPath:
          path: /var/run/secrets/mqtt/edge-agent
          type: DirectoryOrCreate
      - name: edge-privatekey
        hostPath:
          path: /var/run/secrets/pem/edge-agent-private-key
          type: DirectoryOrCreate
      - name: edge-publickey
        hostPath:
          path: /var/run/secrets/pem/edge-agent-public-key
          type: DirectoryOrCreate
      - name: mqtt-volume
        hostPath:
          path: /var/ability/mosquitto
          type: DirectoryOrCreate
      - name: mqtt-auth-volume
        hostPath:
          path: /var/ability/mqtt_auth
          type: DirectoryOrCreate


      imagePullSecrets:
      - name: abilitycr
```

**Comment by BenTheElder** [comment]
> hostPath type check failed: /var/ability/modules is not a directory

This is your problem.
When you pick the `Directory` hostpath type it must exist on the host and must be a directory, as your responsibility.
You can instead use one of the other types like some of your other volumes, or double check that you've created this as expected.

**Comment by srinicrick65** [comment]
Hi Ben the directory is available in my host. I have fixed this issue . This issue is there in **v1.20.4-k3s**1 and the issue is fixed in **v1.20.5-k3s1** but k3d (k3d version v4.4.0) is using v1.20.4-k3s1 . can you upgrade you stable version of k3s to v1.20.5-k3s1 or apply the volume mount fix

**Comment by BenTheElder** [comment]
K3s is not part of the Kubernetes project. We have nothing to do with it and I have no idea when it will be fixed. K3s is a fork. Please file k3s bugs with k3s.
